File: app/src/main/java/com/qinshou/rcvbaseadapterdemo/itemview/BaseItemView.java
package com.qinshou.rcvbaseadapterdemo.itemview;

import android.content.Context;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;

import com.qinshou.rcvbaseadapterdemo.holder.BaseViewHolder;

import java.lang.reflect.ParameterizedType;
import java.lang.reflect.Type;

/**
 * Description:RecyclerView 单数据对多类型布局，多数据对多类型布局的适配器的 BaseItemView
 * Created by 禽兽先生
 * Created on 2018/4/9
 */

public abstract class BaseItemView<T> {
    private Context context;
    private int layoutId;

    public BaseItemView(Context context, int layoutId) {
        this.context = context;
        this.layoutId = layoutId;
    }

    public BaseViewHolder onCreateViewHolder(ViewGroup parent) {
        View itemView = LayoutInflater.from(context).inflate(layoutId, parent, false);
        return new BaseViewHolder(context, itemView);
    }

    /**
     * Description:     子类可以覆蓋此方法决定引用该子布局的时机
     * Date:2018/8/6
     *
     * @param item     该position对应的数据
     * @param position position
     * @return 是否属于子布局
     */
    public boolean isForViewType(T item, int position) {
        Type type = getClass().getGenericSuperclass();
        ParameterizedType parameterizedType = (ParameterizedType) type;
        Class clazz = (Class) parameterizedType.getActualTypeArguments()[0];
        return item.getClass() == clazz;
    }

    /**
     * Description:绑定 UI 和数据
     * Date:2018/8/6
     */
    public abstract void bindViewHolder(BaseViewHolder holder, T itemData, int position);
}

File: rcvbaseadapter/src/main/java/com/qinshou/rcv/itemview/BaseItemView.kt
package com.qinshou.rcv.itemview

import android.content.Context
import android.view.LayoutInflater
import android.view.ViewGroup
import com.qinshou.rcv.adapter.RcvBaseAdapter
import com.qinshou.rcv.viewholder.BaseViewHolder
import java.lang.reflect.ParameterizedType
import java.lang.reflect.Type

/**
 * Author: MrQinshou
 * Email:cqflqinhao@126.com
 * Date: 2021/4/14 13:47
 * Description:The basic of item view,if want to ono-to-many or many-to-many layout,you may use it.
 */
abstract class BaseItemView<T>(val context: Context, private val layoutId: Int, val rcvBaseAdapter: RcvBaseAdapter<Any>) {

    abstract fun bindViewHolder(holder: BaseViewHolder, itemData: T?, position: Int)

    fun onCreateViewHolder(parent: ViewGroup?): BaseViewHolder {
        return BaseViewHolder(context, LayoutInflater.from(context).inflate(layoutId, parent, false))
    }

    open fun isForViewType(item: Any?, position: Int): Boolean {
        if (item == null) {
            return false
        }
        val genericSuperclass = javaClass.genericSuperclass
        if (genericSuperclass !is ParameterizedType) {
            return false
        }
        if (genericSuperclass.actualTypeArguments.isEmpty()) {
            return false
        }
        val type: Type = genericSuperclass.actualTypeArguments[0]
        return item::class.java == type
    }
}